jinja2/filters.py

```diff
@@ -977,10 +984,17 @@ def do_map(*args, **kwargs):
 
     .. versionadded:: 2.7
     """
+    try:
+        default = kwargs['default']
+    except KeyError:
+        default = ''
     seq, func = prepare_map(args, kwargs)
     if seq:
         for item in seq:
-            yield func(item)
+            if func(item):
```

Comment:
This should be a try catch block. Right now, it's replacing falsey values with the default, which it shouldn't do, but it will still fail if the attribute doesn't exist, which is what the issue was about. Also, it causes the function to be called twice.

Reply:
See my below commits. Refactored so the function no longer needs to be called twice, also now checking for undefined type instead of falsey values. Added tests as well.